### PR TITLE
bpo-39303: [WIP] Refactor cmd module

### PR DESCRIPTION
**Pull request by dan1994**
Hello everyone!

As stated in the issue, I'm new to contributing to open source, but would love to contribute to python!

Any comments about styling, testing, or even things like breaking this issue into smaller issues would be highly appreciated!

https://bugs.python.org/issue39303

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@dan1994

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Review by corona10 — CHANGES_REQUESTED** [review]
https://docs.python.org/3.9/library/cmd.html

Thanks for the contribution, however, this change will affect the public APIs of the cmd module.
This kind of work should be processed after the discussion with core developers.

**Comment by dan1994** [comment]
Where should discussion be held, and how would I contact the core developers?
In any case, at least for the time being, I preserve existing APIs, and only partition into functions (although in my opinion there are some changes that will benefit the module).

File: Lib/cmd.py [hunk 1]
```diff
@@ -84,14 +84,8 @@ def __init__(self, completekey='tab', stdin=None, stdout=None):
...
-        if stdin is not None:
-            self.stdin = stdin
-        else:
-            self.stdin = sys.stdin
-        if stdout is not None:
-            self.stdout = stdout
-        else:
-            self.stdout = sys.stdout
+        self.stdin = stdin if stdin is not None else sys.stdin
+        self.stdout = stdout if stdout is not None else sys.stdout
```
Comment (0xcurb) [inline Lib/cmd.py]: I think `or` would work here `self.stdin = stdin or sys.stdin`

**Comment by dan1994** [comment]
Meanwhile I'll spill some of my thoughts and questions:

So far I've been adding changes in ways that should not break the existing API (hopefully). I've been careful not to change the behavior, and focused on giving meaningful names and dividing into functions so the module will be more readable.

I do have a few things I would very much like to change, but would love some professional opinion about:
- The Single Responsibility Principle dictates that this class should be split to more classes.  The responsibilities I identify and think should have their own classes are:
    - Line parsing
    - Completion
    - Help printing
- Given that I split into classes I'm not sure how and where should I place them. 
    - Should they be in separate files to leave the `cmd.py` file cleaner (this will practically create more modules)? 
    - Should they be in the same file but not listed in `__all__` (leaving `cmd.py` pretty bloated)? 
    - Should they  be inner classes of `cmd` (again leaving `cmd.py` pretty bloated)?
- In the current situation, `!` is an alias to `shell`, but there's no default implementation to `do_shell`. This causes an ugly edge case. I would like to provide a default `do_shell` implementation that prints that there's no shell implementation.
- In general, a nice feature would be to add aliases to commands, but this should probably be handled in a separate issue.
- I wonder if there's a reason (or if it will make any difference) for variables such as `prompt`, `intro` and so forth not to be instantiated in `__init__` as members.
- I did not yet refactor the `complete` function, because I can't test it without `readline` (that doesn't come with `python` by default). Even if I installed `readline` locally, I can't add tests, since the CI (and other people) won't have `readline` installed. Is there a way around this?

**Comment by csabella** [comment]
@dan1994, thank you for your contribution and for your interest in improving CPython.  We do appreciate your enthusiasm, however there is a very high bar for refactoring code in this repo and we generally don't do it based on the benefit to cost.  The `cmd` module has existed for over 20 years and hasn't had any new patches applied since 2013.  There also isn't a current core developer who is a code owner for the module to take on this change.  Please take a look at [this talk](https://t.co/yDSvZwjddk) or [this talk](https://www.youtube.com/watch?v=voXVTjwnn-U) to get a better understanding of the reasons for that.

I'm going to close this, but please don't let that discourage you from participating in the future.  If you take a look at those videos (and others), you'll see there are a lot of ways in which you can help out and contribute.

**Comment by dan1994** [comment]
Thank you for your response. I hope to find other opportunities to contribute, and in any case, this has been a nice exercise for me, even if it ended a bit prematurely.

**Pull request closed without merging**